Newsletters: add 154 (2021-06-23) (#592)
bitcoinops/bitcoinops.github.io: _includes/specials/taproot/00-bech32m.md

Starting at block {{site.trb}}, expected in November, Bitcoin users will be
able to safely receive payments to taproot  addresses.  Given the
user enthusiasm for taproot and the five months that wallet
developers have to implement support for it, Optech expects there to be
several popular wallets that will allow their users to generate taproot
addresses at the earliest possible moment.

That means any other wallet or service that sends bitcoins to

> user-provided addresses needs to be able to send to taprot addresses by

Reviewer: ```suggestion
user-provided addresses needs to be able to send to taproot addresses by
```

Author: You think this is optional?  :-D

Reviewer: Oops, no not that one.

block {{site.trb}} or risk confusing and disappointing its users.  Pay
to TapRoot (P2TR) addresses use [bech32m][topic bech32] as specified in [BIP350][], which
is slightly different than [BIP173][]'s bech32 algorithm used for segwit

> v0 P2WPKH and P2WSH addresses.  The slight difference is that the
> checksum function uses a different constant, with bech32m using
> `0x2bc830a3` instead of bech32's `0x01`.

Reviewer: "different, difference, different" in three lines. Perhaps:

```suggestion
v0 P2WPKH and P2WSH addresses.  Bech32m uses the constant
`0x2bc830a3` instead of bech32's `0x01` in the checksum function.
```

## Preparing for taproot #1: bech32 sending support

> {% assign trb = "709,632" %}{% comment %}<!-- stop me from mistyping the number -->{% endcomment %}

Reviewer 2: Perhaps make this a site-wide alias? I'm sure we'll be referring to 709,632 in many places over the next few months.

*The first in a weekly series about how developers and service providers
can prepare for the upcoming activation of taproot at block height
{{trb}}.*

Starting at block {{trb}}, expected in November, Bitcoin users will be
able to safely receive payments to taproot  addresses.  Given the
current enthusiasm for taproot and the five months that wallet
developers have to implement support for it, Optech expects there to be
several popular wallets that will allow their users to generate taproot

> addresses within days of block {{trb}}, if not hours.

Author: I think "allow" is the correct word here; I'm not sure what grammar rule that is, but my native speaker intuition is insisting on it.

That means any other wallet or service that sends bitcoins to

> user-provided addresses needs to be able to Pay To TapRoot (P2TR) by

Reviewer 2: Using Pay To Taproot/P2TR as a verb/phrase is a little confusing, and I think we should discourage it in general. We don't say "the sender paid to script hash" but rather "the sender/wallet created a P2SH output" or "the receiver generated a P2SH address".

Here, I'd suggest changing the sentence to read "needs to be able to send to P2TR (Pay To TapRoot) addresses" or similar.

> block {{trb}} or risk confusing and disappointing its users.  Taproot

Author: I think this change is a good idea on its own, but after the changes from the previous suggestion of dropping the block number, the text "At that point" becomes an orphaned pointer.

addresses use [bech32m][topic bech32] as specified in [BIP350][], which
is slightly different than [BIP173][]'s bech32 algorithm used for segwit
v0 P2WPKH and P2WSH addresses.  The slight difference is that the
checksum function uses a different constant, with bech32m using
`0x2bc830a3` instead of bech32's `0x01`.

> Changing that single constant in your code will provide you with the

Reviewer 3: Narrative shift from referring to wallets in the third person in the previous paragraphs to the second person ("you")...

ability to verify bech32m checksums, but you still need to use the original constant for existing
P2WPKH and P2WSH addresses.  Your code needs to decode the address
without verifying the checksum, determine whether it uses v0 segwit (bech32) or
v1+ segwit (bech32m), and then validate the checksum with the appropriate
constant.  For examples, see the [PR][bech32#56] that updated the bech32
reference implementations for C, C++, JS, and Python.  If your code already
uses the reference libraries, you can update to the latest code from
that repository, although note that some of the APIs have slight
changes.  BIP350 and the reference implementations provide test vectors
that you are encouraged to use.

Although *receiving* payments to taproot addresses won't be safe until
block {{trb}}, *sending* payments should not cause any problems for the
sender.  Bitcoin Core has supported relaying and mining transactions
with taproot-paying outputs since version 0.19 (released November 2019).

> Optech encourages developers of wallets and services to implement

Reviewer 3: ...and then from "you" to back to the third person: "developers of wallets"
